cluster/gce/gci/configure.sh

```diff
@@ -225,9 +225,105 @@ function install-kube-binary-config {
+  set +e
+  local -r max_attempts=5
+  local -i attempt_num=1
+  until timeout 30 docker load -i "${img}"; do
+    if [[ "${attempt_num}" == "${max_attempts}" ]]; then
+      echo "Fail to load docker image file ${img} after ${max_attempts} retries. Exit!!"
+      exit 1
+    else
+      attempt_num=$((attempt_num+1))
+      sleep 5
+    fi
+  done
+  # Re-enable errexit.
+  set -e
+}
+
+# Loads kube-system docker images. It is better to do it before starting kubelet,
+# as kubelet will restart docker daemon, which may interfere with loading images.
+function load-docker-images {
+  echo "Start loading kube-system docker images"
+  local -r img_dir="${KUBE_HOME}/kube-docker-files"
+  if [[ "${KUBERNETES_MASTER:-}" == "true" ]]; then
+    try-load-docker-image "${img_dir}/kube-apiserver.tar"
+    try-load-docker-image "${img_dir}/kube-controller-manager.tar"
+    try-load-docker-image "${img_dir}/kube-scheduler.tar"
+  else
+    try-load-docker-image "${img_dir}/kube-proxy.tar"
+  fi
+}
+
+# Downloads kubernetes binaries and kube-system manifest tarball, unpacks them,
+# and places them into suitable directories. Files are placed in /home/kubernetes.
+function install-kube-binary-config {
+  cd "${KUBE_HOME}"
+  local -r server_binary_tar_urls=( $(split-commas "${SERVER_BINARY_TAR_URL}") )
+  local -r server_binary_tar="${server_binary_tar_urls[0]##*/}"
+  if [[ -n "${SERVER_BINARY_TAR_HASH:-}" ]]; then
+    local -r server_binary_tar_hash="${SERVER_BINARY_TAR_HASH}"
+  else
+    echo "Downloading binary release sha1 (not found in env)"
+    download-or-bust "" "${server_binary_tar_urls[@]/.tar.gz/.tar.gz.sha1}"
+    local -r server_binary_tar_hash=$(cat "${server_binary_tar}.sha1")
+  fi
+
+  preload=$(is-preloaded "${server_binary_tar}" "${server_binary_tar_hash}")
+  if [[ preload -eq 0 ]]; then
+    echo "${server_binary_tar} is preloaded."
+  else
+    echo "Downloading binary release tar"
+    download-or-bust "${server_binary_tar_hash}" "${server_binary_tar_urls[@]}"
+    tar xzf "${KUBE_HOME}/${server_binary_tar}" -C "${KUBE_HOME}" --overwrite
+    # Copy docker_tag and image files to ${KUBE_HOME}/kube-docker-files.
+    src_dir="${KUBE_HOME}/kubernetes/server/bin"
+    dst_dir="${KUBE_HOME}/kube-docker-files"
+    mkdir -p "${dst_dir}"
+    cp "${src_dir}/"*.docker_tag "${dst_dir}"
+    if [[ "${KUBERNETES_MASTER:-}" == "false" ]]; then
+      cp "${src_dir}/kube-proxy.tar" "${dst_dir}"
+    else
+      cp "${src_dir}/kube-apiserver.tar" "${dst_dir}"
+      cp "${src_dir}/kube-controller-manager.tar" "${dst_dir}"
+      cp "${src_dir}/kube-scheduler.tar" "${dst_dir}"
+      cp -r "${KUBE_HOME}/kubernetes/addons" "${dst_dir}"
+    fi
+    load-docker-images
+    mv "${src_dir}/kubelet" "${KUBE_BIN}"
+    mv "${src_dir}/kubectl" "${KUBE_BIN}"
+
+    mv "${KUBE_HOME}/kubernetes/LICENSES" "${KUBE_HOME}"
+    mv "${KUBE_HOME}/kubernetes/kubernetes-src.tar.gz" "${KUBE_HOME}"
+  fi
+
+  if [[ "${KUBERNETES_MASTER:-}" == "false" ]] && \
```

Comment:
NPD binaries are downloaded and installed separately (not together with k8s release). The previous location of this code doesn't make much sense to me. I'm moving the npd installation to its own function to make the code (especially the preload check) cleaner.